src/common/execute.ts

```diff
@@ -0,0 +1,36 @@
+import { FailureReason, IBasePolicyOptions } from '../Policy';
+
+export type FailureOrSuccess<R> = FailureReason<R> | { success: R };
+
+export const returnOrThrow = <R>(failure: FailureOrSuccess<R>) => {
+  if ('error' in failure) {
+    throw failure.error;
+  }
+
+  if ('success' in failure) {
+    return failure.success;
+  }
+
+  return failure.value;
+};
+
+export const execute = async <T extends any[], R>(
+  options: Readonly<IBasePolicyOptions<R>>,
+  fn: (...args: T) => Promise<R> | R,
```

Comment:
Why not have the caller wrap the args instead?
eg. `execute(options: Readonly<IBasePolicyOptions<R>>, fn: () => Promise<R> | R)`, that way you can avoid the generic.
Also capturing a vararg is probably not much cheaper than creating another closure, benchmark welcome ofc :)

Reply:
For example, `setTimeout` techically allows you to pass a vararg to your callback, almost nobody does that anymore though :)

Reply:
Huh, I didn't expect it to be the case, but it looks like spreading is indeed faster. I guess they optimized that!

```ts
const xor = (a, b) => a ^ b;
const varargs = (fn, ...args) => fn(...args);
const closure = fn => fn();

let x = 0;
bench('varargs', () => x = varargs(xor, x, x * 2));
bench('closure', () => x = closure(() => xor(x, x * 2)));

/*
     350,218,743 op/s » varargs
     192,205,296 op/s » closure
```